### PR TITLE
feat: support always-auth

### LINKED ISSUE #600: Only allow logged in users to install packages
Is there a mode where the registry can be hosted publicly but no APIs are public to anonymous users.

I want only logged in users to be able to install packages.

### PR DESCRIPTION
**Pull request by fengmk2**
Add `alwaysAuth: true` to config and set npm cli config.

Use 'http://registry.sample.com' for example, set npm config below:

    $ npm config set '//registry.sample.com/:always-auth=true'
    $ npm login --registry=http://registry.sample.com
    $ npm install mocha --registry=http://registry.sample.com

Fixes #600

### CODE REVIEW

**Comment by KidkArolis** [comment]
Seems to work! That was fast! Thank you.

Does this basically close down 100% of the endpoints to non logged in users?

**Comment by KidkArolis** [comment]
What about registration (npm login / npm adduser) - is it possible to disable that?

File: middleware/auth.js [hunk 1]
```diff
@@ -62,3 +63,19 @@ module.exports = function () {
     yield* next;
   };
 };
+
+function* unauthorized(next) {
+  if (!config.alwaysAuth || this.method !== 'GET') {
```
Comment (fengmk2) [inline middleware/auth.js]: @KidkArolis registration is `PUT` request, it will be pass to login controller.

**Comment by dead-horse** [comment]
awesome

**Comment by fengmk2** [comment]
+ cnpmjs.org@2.0.0-rc.14

**Pull request merged**